x-pack/plugins/security_solution/server/lib/app_features_service/app_features_service.ts

```diff
@@ -99,4 +100,62 @@ export class AppFeaturesService {
     }
     return this.appFeatures.has(appFeatureKey);
   }
+
+  public getApiActionName = (apiPrivilege: string) => `api:${APP_ID}-${apiPrivilege}`;
+
+  public isActionRegistered(action: string) {
+    return (
+      this.securityAppFeatures.isActionRegistered(action) ||
+      this.casesAppFeatures.isActionRegistered(action) ||
+      this.securityAssistantAppFeatures.isActionRegistered(action)
+    );
+  }
+
+  public isApiPrivilegeEnabled(apiPrivilege: string) {
+    return this.isActionRegistered(this.getApiActionName(apiPrivilege));
+  }
+
+  public registerApiAccessControl(http: HttpServiceSetup) {
+    // The `securitySolutionAppFeature:` prefix is used for AppFeature based control.
+    // Should be used only by routes that do not need RBAC, only direct appFeature control.
+    const APP_FEATURE_TAG_PREFIX = 'securitySolutionAppFeature:';
+    // The "access:securitySolution-" prefix is used for API action based control.
+    // Should be used by routes that need RBAC, extending the `access:` role privilege check from the security plugin.
+    // An additional check is performed to ensure the privilege has been registered by the appFeature service,
+    // preventing full access (`*`) roles, such as superuser, from bypassing appFeature controls.
+    const API_ACTION_TAG_PREFIX = `access:${APP_ID}-`;
+
+    http.registerOnPostAuth((request, response, toolkit) => {
```

Comment:
__nit:__ this hook will be called for every request to Kibana, so it has quite a high performance bar. You might want to minimize the number of loops you go through in the tags list. Maybe something along these lines:
```ts
      for (const tag of request.route.options.tags) {
        let isEnabled = true;
        if (tag.startsWith(APP_FEATURE_TAG_PREFIX)) {
          isEnabled = this.isEnabled(
            tag.substring(APP_FEATURE_TAG_PREFIX.length) as AppFeatureKeyType
          );
        } else if (tag.startsWith(API_ACTION_TAG_PREFIX)) {
          isEnabled = this.isApiPrivilegeEnabled(tag.substring(API_ACTION_TAG_PREFIX.length));
        }

        if (!isEnabled) {
          this.logger.warn(
            `Accessing disabled route "${request.url.pathname}${request.url.search}": responding with 404`
          );
          return response.notFound();
        }
      }
```